olmo/train.py

```diff
@@ -807,14 +856,35 @@ def train_batch(self, batch: Dict[str, Any]) -> Tuple[torch.Tensor, Optional[tor
                         assert z_batch_loss is not None
                         z_batch_loss += z_loss.detach()
 
+                if self.model.config.block_type == BlockType.moe:
+                    if self.model.config.moe_zloss_weight is not None:
+                        lb_loss, moe_z_loss = batched_load_balancing_loss(self.moe_args)
+                        lb_loss = lb_loss / len(micro_batches)
+                        moe_z_loss = moe_z_loss / len(micro_batches)
+                    else:
+                        lb_loss = batched_load_balancing_loss(self.moe_args) / len(micro_batches)
+                    if self.model.config.moe_log_expert_assignment:
+                        if self.model.config.moe_zloss_weight:
+                            tokens_per_expert, _, _ = zip(*get_load_balancing_loss())
+                        else:
+                            tokens_per_expert, _ = zip(*get_load_balancing_loss())
+                        expert_assignments += torch.stack(tokens_per_expert, dim=0).cpu()
+                    clear_load_balancing_loss()
+                    if self.model.config.moe_loss_weight:
+                        loss += lb_loss
+                    if self.model.config.moe_zloss_weight:
+                        loss += moe_z_loss
+                    lb_batch_loss += lb_loss.detach()
+                    moe_z_batch_loss += moe_z_loss.detach()
+
                 # Run backward pass.
                 loss.backward()
 
             # Remove output hooks
             for hook in output_hooks:
                 hook.remove()
 
-        return ce_batch_loss, z_batch_loss
+        return ce_batch_loss, z_batch_loss, lb_batch_loss, moe_z_batch_loss, expert_assignments
```

Comment:
@epwalsh, does the new trainer support all of this stuff? This seems like a lot of extra things.

Reply:
Not directly but I think it could be supported through the callback system.